Feature/choose beanstalk version

timdev opened the pull request:
Allow picking version via configuration.

timdev committed:
Allow selecting beanstlakd version via configuration.  Tested for 1.10, but reset to 1.9 for sake of upstream.

timdev committed:
Updated README.md to show beanstalkd_version default.

ramondelafuente commented on tasks/main.yml:
```diff
@@ -1,15 +1,15 @@
 - name: Install
-  command: make install chdir=/tmp/beanstalkd-1.9/ creates=/usr/local/bin/beanstalkd
+  command: make install chdir=/tmp/beanstalkd-{{ beanstalkd_version}}/ creates=/usr/local/bin/beanstalkd
 
```
Due to the `creates` line, the `make install` command will not run if you've already got 1.9
Not really a big issue, but having the variable creates the impression that this role will upgrade... 

Any thoughts?

timdev replied:
Good point.  Simply removing the `creates` seems ugly.   I'll give it some thought later and see if I can come up with a reasonable fix.

timdev commented:
Having given this a bit more thought, I think removing the `creates` (and changing the final `state=started` to `state=restarted`) is probably the best option after all.

On the plus side, it would be mostly idempotent.  And it's simple.

On the downside, it's imperfect.  The daemon would be restarted even if no upgrade actually happened.  Considering the default of no persistence, that could be bad.

The other option would be using something like the [script module](http://docs.ansible.com/script_module.html) to upload and execute some script that would test for the presence of beanstalkd, check it's version, and then copy and (re)start as needed.

Any preference, or other ideas?

ramondelafuente commented on tasks/main.yml:
```diff
@@ -1,15 +1,15 @@
 - name: Get tarfile
-  command: wget https://github.com/kr/beanstalkd/archive/v1.9.tar.gz -q -O /tmp/beanstalkd.tar.gz creates=/tmp/beanstalkd.tar.gz
+  command: wget https://github.com/kr/beanstalkd/archive/v{{ beanstalkd_version }}.tar.gz -q -O /tmp/beanstalkd.tar.gz creates=/tmp/beanstalkd.tar.gz
 
```
Also add the version to the tmp-file, otherwise newer versions won't get downloaded

ramondelafuente committed:
Updated script to allow for beanstalkd upgrades

- The default version is now 1.10
- When running this role on a new server, Beanstalkd will be installed and run (as before)
- When running this role with a different version as the installed Beanstalkd, the
existing binary will be removed and the installation procedure will be re-run.
- A handler is included to restart Beanstalkd on upgrade/downgrade

ramondelafuente commented:
I've been playing around with this, and it turned out the `Make install` command also did not honor the upgrading to a newer version. I've put together a working upgrade (and downgrade) version of your patch here:

https://github.com/f500/ansible-beanstalkd/commits/timdev-feature/choose-beanstalk-version

Could you take a look?

timdev committed:
Updated script to make persistence default, and enable system user configuration

- beanstalkd_persistent now defaults to "yes"
- added beanstalkd_user variable
- ensure beanstalkd_user exist
- ensure beanstalkd_user owns the directory beanstalkd_storage (and all contents)
- no longer bothers to download/untar tarball if no version change
- made tasks names more descriptive/less imperative

timdev committed:
Only force service to stop if process/storage owner is observed to change.

timdev commented:
That ended up being a bit more work than I thought, mainly because I decided to allow configuration of the account the daemon runs under.  I hope you don't mind the scope creep, @ramondelafuente 

This PR now makes persistence the default, and allows setting or changing the user the daemon runs as.

Solving that made me realize that changing the storage location should also be handled more gracefully.  Currently, if you want to move the storage location, you'll lose the binlog.  It would be better to detect that change somehow, and try to move the storage directory instead of just creating it.

But that was the case before this PR anyway.  I think this can probably be merged as-is.  Let me know if you need anything else.

ramondelafuente commented:
We've got a couple more roles that _try_ to handle such changes, but it's a headache because you need the old location as well as the new. We mostly just use beanstalkd as-is anyways, so let's wait until we need it or people actually start complaining ;-)

ramondelafuente commented on tasks/main.yml:
```diff
@@ -1,24 +1,49 @@
-  command: make install chdir=/tmp/beanstalkd-1.9/ creates=/usr/local/bin/beanstalkd
+- name: Get tarfile.
+  command: wget https://github.com/kr/beanstalkd/archive/v{{ beanstalkd_version }}.tar.gz -q -O /tmp/beanstalkd-{{ beanstalkd_version }}.tar.gz creates=/tmp/beanstalkd-{{ beanstalkd_version }}.tar.gz
+  when: installed_beanstalkd_version.stdout != beanstalkd_version
```
Did I add this 'when' ? I'd think the 'creates' handles this situation, and the when is overkill. On the other hand, the when skips the task, which should be quicker than the file-existance check...

ramondelafuente commented on tasks/main.yml:
```diff
@@ -1,24 +1,49 @@
-  copy: src=beanstalkd.initd dest=/etc/init.d/beanstalkd mode=0755
+- name: Unarchive.
+  command: tar -zxvf /tmp/beanstalkd-{{ beanstalkd_version }}.tar.gz chdir=/tmp creates=/tmp/beanstalkd-{{ beanstalkd_version }}
+  when: installed_beanstalkd_version.stdout != beanstalkd_version
```
Did I add this 'when' ? I'd think the 'creates' handles this situation, and the when is overkill. On the other hand, the when skips the task, which should be quicker than the file-existance check... (repeated)

ramondelafuente commented on tasks/main.yml:
```diff
@@ -1,24 +1,49 @@
-  template: src=beanstalkd dest=/etc/default/beanstalkd
+- name: Make.
+  command: make chdir=/tmp/beanstalkd-{{ beanstalkd_version}}/ creates=/tmp/beanstalkd-{{ beanstalkd_version}}/beanstalkd
+  when: installed_beanstalkd_version.stdout != beanstalkd_version
```
Did I add this 'when' ? I'd think the 'creates' handles this situation, and the when is overkill. On the other hand, the when skips the task, which should be quicker than the file-existance check... (repeated)

timdev replied:
I added those 'when's while testing the owner-juggling stuff.  No reason to not skip them if the version isn't changing.

ramondelafuente commented:
Oh, and I don't mind the scope-creep - thank _you_ for all the work on this!

timdev commented:
I agree it's not worth worrying about the storage location changing.  Probably best solved with a note in the documentation.

timdev committed:
Updated docs.
